File: python/sdk/src/t0_provider_sdk/provider/validate.py
"""Public helper for validating provider responses against buf.validate rules.

Providers call ``validate(resp)`` before returning to surface validation
failures in the developer's own call frame. On success the input message is
returned unchanged; on failure a ``ConnectError(Code.INTERNAL, ...)`` is
raised with the same wording the SDK's response-validation interceptor
emits, so propagating the error preserves the on-wire shape.

Go equivalent: ``provider.Validate[T]`` (planned).
"""

from __future__ import annotations

from typing import TypeVar

import protovalidate
from connectrpc.code import Code
from connectrpc.errors import ConnectError
from google.protobuf.message import Message

T = TypeVar("T", bound=Message)

# Module-level validator instance, reused by both the helper and the
# response-validation interceptor. ``protovalidate.Validator`` is safe to
# share across threads/tasks.
_validator: protovalidate.Validator | None = None


def _get_validator() -> protovalidate.Validator:
    """Return the shared validator, constructing it on first use."""
    global _validator
    if _validator is None:
        _validator = protovalidate.Validator()
    return _validator


def validate(msg: T) -> T:
    """Validate ``msg`` against its proto rules.

    Args:
        msg: The response message to validate.

    Returns:
        The same ``msg`` instance on success.

    Raises:
        ConnectError: ``Code.INTERNAL`` with message
            ``"response validation failed: <details>"`` when ``msg`` violates
            its proto rules. This matches the SDK response-validation
            interceptor so propagating the error keeps wire behavior
            identical.
    """
    try:
        _get_validator().validate(msg)
    except protovalidate.ValidationError as e:
        raise ConnectError(Code.INTERNAL, f"response validation failed: {e}") from e
    return msg

File: go/provider/validate.go
// Package provider exposes helpers used by provider services. Validate runs the
// same protovalidate rules that the response-validation interceptor enforces,
// but inside the handler's own call frame so the developer can see and react to
// failures before they reach the wire.
package provider

import (
	"errors"
	"fmt"
	"sync"

	"buf.build/go/protovalidate"
	"connectrpc.com/connect"
	"google.golang.org/protobuf/proto"
)

// validator is shared by every Validate call. protovalidate.New caches compiled
// rule programs per message descriptor, so re-using one instance avoids
// recompiling on each call. Construction is lazy because protovalidate.New can
// fail at process startup and we don't want to penalize callers who never use
// the helper.
var validator = sync.OnceValue(func() protovalidate.Validator {
	// protovalidate.GlobalValidator falls back to a lazily-initialised default
	// instance. Returning it here keeps the helper and the connect validate
	// interceptor pointed at the same cache.
	return protovalidate.GlobalValidator
})

// Validate runs protovalidate on msg and returns it unchanged on success.
// On failure it returns the zero value of T and an error whose message is
// prefixed "response validation failed: " so it surfaces consistently with the
// interceptor safety net.
//
// Usage:
//
//	resp, err := provider.Validate(&payment.PayOutResponse{ /* ... */ })
//	if err != nil {
//	    return nil, err
//	}
//	return connect.NewResponse(resp), nil
//
// Letting the returned error propagate keeps the wire response bit-identical to
// what the interceptor would have produced. Catching it lets the handler
// convert the failure into a domain-level error (for example, the Failed arm
// of a oneof result) before responding.
func Validate[T proto.Message](msg T) (T, error) {
	if err := validator().Validate(msg); err != nil {
		var zero T
		return zero, connect.NewError(connect.CodeInternal,
			fmt.Errorf("response validation failed: %w", err))
	}
	return msg, nil
}

// asValidationError extracts a *protovalidate.ValidationError from err if one
// is wrapped inside, returning nil otherwise. Exposed unexported so the
// interceptor and tests can share the same extraction logic.
func asValidationError(err error) *protovalidate.ValidationError {
	var ve *protovalidate.ValidationError
	if errors.As(err, &ve) {
		return ve
	}
	return nil
}
